mba6x_bl: init at 2016-02-12

simonvandel opened the pull request:
###### Things done:
- [x] Tested using sandboxing (`nix-build --option build-use-chroot true` or [nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS)
- [x] Built on platform(s): NixOS
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

This driver fixes backlight after suspend/resume on Macbook Air 6,1
and 6,2.

Added a workaround for https://github.com/patjak/mba6x_bl/issues/43 in
form of a systemd service.

Has been working great on my machine for at least a month now.

simonvandel committed:
mba6x_bl: init at 2016-02-12

This driver fixes backlight after suspend/resume on Macbook Air 6,1
and 6,2.

Added a workaround for https://github.com/patjak/mba6x_bl/issues/43 in
form of a systemd service.

Has been working great on my machine for at least a month now.

joachifm commented on nixos/modules/hardware/video/backlight/mba6x_bl.nix:
```diff
@@ -0,0 +1,27 @@
+    services.mba6x_bl = {
+      enable = mkEnableOption "mba6x_bl";
+    };
+  };
```
This will appear somewhat cryptic to readers of the man page, is it possible to provide a more descriptive name/docs?

simonvandel commented:
Oops, i screwed up. Let med restore this..

simonvandel commented:
@joachifm : I changed the description based on your feedback.

joachifm commented:
Great. I'd like to get some feedback from others before merging this, though. Feel free to ping me if it ends up lingering for too long.

hrdinka commented on nixos/modules/hardware/video/backlight/mba6x_bl.nix:
```diff
@@ -0,0 +1,27 @@
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        Type = "oneshot";
+        ExecStart = "/bin/sh -c \"${modprobe} -r mba6x_bl && ${modprobe} mba6x_bl\"";
```
Please use `script` instead of calling `sh` yourself. It's documented over here: https://nixos.org/nixos/manual/options.html#opt-systemd.services._name_.script

hrdinka commented on nixos/modules/hardware/video/backlight/mba6x_bl.nix:
```diff
@@ -0,0 +1,27 @@
+  config = mkIf cfg.enable {
+    boot.extraModulePackages = [ kernel.mba6x_bl ];
+    systemd.services.mba6x_blWorkaround = {
+      description = "Remove and reload mba6x to workaround no brightness bug (#43)";
```
It's not clear to which project Issue 43 belongs. Maybe you can add a hint where to find it.

hrdinka commented on pkgs/os-specific/linux/mba6x_bl/default.nix:
```diff
@@ -0,0 +1,33 @@
+with stdenv.lib;
+
+let version = "2016-02-12";
+    pkgName = "mba6x_bl";
```
Just nitpicking: these variables are only used once, so there is no need to extra define them here.

hrdinka commented:
Adding a whole module for a temporary workaround feels kinda wrong. Does this bug also occur when the module isn't auto loaded (by blacklisting it) but loaded manually or does it always need to be loaded twice to work?

simonvandel commented:
@hrdinka Thanks for your feedback. I will add fixes soon.

I don't know if it can be justified to make it a module. The driver can be loaded by users by setting `boot.extraModulePackages = [ kernel.mba6x_bl ];` in their config file. I feel it is nicer for a module to do this. The workaround was easily added as a systemd unit to the module, making it work as expected without exposing this to the user.

hrdinka commented:
Ok I see. So it's not only for the workaround. Then it might fit better as an extra option below `programs.light` (@puffnfresh) something like `programs.light.mba6xSupport`. @joachifm what's your opinion on that?

simonvandel commented:
It seems weird to relate the mba6x driver to the light utility. How do they relate other than being in the domain of background light?

joachifm commented:
Hm, what about the `hardware.` option namespace?

hrdinka commented:
@joachifm that was my second idea. This option shouldn't be placed in services.

@simonvandel yes you are right. This was a fast fire. Technically it does not refer to the light program at all. I just like the idea of having it somewhere where it belongs instead of adding another top level option for a specific device that might be unnecessary in future. Something like:

`hardware.backlight.enable` # enable generic backlight control
`hardware.backlight.macbookAir6xSupport` # enables specific device drivers
`hardware.backlight.samsungWhateverSupport` # or workarounds
...

`programs.light` seemed natural to me because it is our only backlight control related option in nixos and might be the right backend for a more general backlight option.

simonvandel commented:
> Hm, what about the hardware. option namespace?

This seems more appropriate.

simonvandel commented:
> Something like:
> 
> hardware.backlight.enable # enable generic backlight control
> hardware.backlight.macbookAir6xSupport # enables specific device drivers
> hardware.backlight.samsungWhateverSupport # or workarounds

It would be nice to have a general framework for these kind of hardware specific packages. Another alternative is to leave the actual driver be in nixpkgs, but drop the module support. The use of the driver could then be documented in something like [https://github.com/ehmry/nixos-hardware](https://github.com/ehmry/nixos-hardware).

hrdinka commented:
Or a more general hardware support and fixing scheme:

`hardware.support.macbookAir6x` # enables macbook air 6,x related drivers/fixs
`hardware.support.dellXY`
`hardware.support.hpXY`

hrdinka commented:
@simonvandel thank's for the link, didn't know that. That's actually what my last propose would do under the hood.

hrdinka commented:
My previous idea is a bad one. An entry for every device out there isn't practical and is absolutely impossible to maintain. I guess for your cause your last post shows the best solution. Just keep the package and add some documentation about its existence and the workaround. The bug itself is hopefully fixed soon upstream.

joachifm commented:
Suggestion: if the driver is useful on its own, submit a separate PR for it, then we can figure out how to best expose it to users later.

simonvandel commented:
I updated this PR to only include the packaging of the driver, and implemented the feedback received. Is this merge-able?

I will add some documentation to https://github.com/ehmry/nixos-hardware when I get the time.

joachifm commented:
It looks to be in keeping with similar packages already in the tree. If it's useful to you, it's sure to be useful to someone else. Thank you for your efforts.

huacayacauh commented:
Hello
May I ask how to use this package ? I have the exact problem it is supposed to fix (I am running nixos 15.09 on a macbook air 6.2), but can't manage to find out how to include it in my configuration...
Thank you for the work!

joachifm commented:
@huacayacauh did you try https://github.com/NixOS/nixpkgs/pull/13688#issuecomment-192967310

huacayacauh commented:
@joachifm thanks. Yes, in this case nixos-rebuild tells me that kernel variable is undefined. I don't know what I should substitute kernel for, my tries gave "attribute mba6x_bl missing".

joachifm commented:
I think you'd need to refer to `boot.kernelPackages.mba6x_bl` (untested)

huacayacauh commented:
@joachifm 
`boot.kernelPackages.mba6x_bl` gives `undefined variable ‘boot’`
`config.boot.kernelPackages.mba6x_bl` gives `attribute ‘mba6x_bl’ missing`

joachifm commented:
Oh, I see you're running 15.09; this is not available there.

joachifm commented:
It won't be in a release channel for a while. The earliest you'll get it is when the nixos-unstable channel becomes unstuck. Alternatively, you can clone NixOS/nixpkgs-channels at the release branch you want and cherry-pick the commit onto it. I won't go into details here, see the manual.

huacayacauh commented:
@joachifm Ok, thank you very much for the quick answer.
